File: clipsync-win/ClipSync.Core/Settings/AppSettings.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClipSync.Settings;

/// A discovered-but-untrusted device the user chose not to see in the
/// peer list. The name is whatever it advertised when hidden, kept so
/// the settings list can say more than a hash.
public sealed record HiddenPeer(string DidHex, string Name);

/// User preferences, stored as plain JSON in %LOCALAPPDATA%\ClipSync.
/// Unlike the trust store this is not secret, and being hand-editable is
/// a feature. A corrupt file degrades to defaults rather than throwing,
/// matching TrustStore.Load().
public sealed class AppSettings
{
    private const int CurrentVersion = 1;

    private readonly string _path;
    private readonly List<AppIdentity> _excluded;
    private readonly List<string> _pausedPeers;
    private readonly List<HiddenPeer> _hidden;
    private readonly object _lock = new();

    private AppSettings(string path, List<AppIdentity> excluded, List<string> pausedPeers,
                        List<HiddenPeer> hidden)
    {
        _path = path;
        _excluded = excluded;
        _pausedPeers = pausedPeers;
        _hidden = hidden;
    }

    public static string DefaultPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "ClipSync", "settings.json");

    public static AppSettings Load() => Load(DefaultPath);

    public static AppSettings Load(string path)
    {
        var list = new List<AppIdentity>();
        var paused = new List<string>();
        var hidden = new List<HiddenPeer>();
        try
        {
            if (File.Exists(path))
            {
                var model = JsonSerializer.Deserialize<FileModel>(File.ReadAllText(path), JsonOptions);
                foreach (var e in model?.ExcludedApps ?? new List<FileModel.Entry>())
                {
                    if (TryParse(e, out var id)) list.Add(id);
                }
                foreach (var did in model?.PausedPeers ?? new List<string>())
                {
                    if (Normalise(did) is { } key && !paused.Contains(key)) paused.Add(key);
                }
                foreach (var h in model?.HiddenPeers ?? new List<FileModel.HiddenEntry>())
                {
                    if (Normalise(h.Did) is { } key && hidden.TrueForAll(x => x.DidHex != key))
                        hidden.Add(new HiddenPeer(key, h.Name ?? key[..Math.Min(8, key.Length)]));
                }
            }
        }
        catch (Exception ex)
        {
            // Corrupt or unreadable: start empty rather than crashing the app.
            Security.Log.Write($"AppSettings: could not read {path}: {ex.GetType().Name}; using defaults");
            list.Clear();
            paused.Clear();
            hidden.Clear();
        }
        return new AppSettings(path, list, paused, hidden);
    }

    public IReadOnlyList<AppIdentity> Excluded
    {
        get { lock (_lock) return _excluded.ToList(); }
    }

    public bool IsExcluded(AppIdentity app)
    {
        lock (_lock) return _excluded.Contains(app);
    }

    public void Add(AppIdentity app)
    {
        lock (_lock)
        {
            if (_excluded.Contains(app)) return;
            _excluded.Add(app);
            Persist();
        }
    }

    public void Remove(AppIdentity app)
    {
        lock (_lock)
        {
            if (_excluded.RemoveAll(e => e.Equals(app)) == 0) return;
            Persist();
        }
    }

    /// Peers this device does not send to, by lowercase DID hex.
    ///
    /// Only the per-peer pause lives here. A global pause is deliberately not
    /// persisted: it means "not right now", and a restart that silently left
    /// syncing off would be a bad surprise.
    public IReadOnlyList<string> PausedPeers
    {
        get { lock (_lock) return _pausedPeers.ToList(); }
    }

    public bool IsPeerPaused(string didHex)
    {
        if (Normalise(didHex) is not { } key) return false;
        lock (_lock) return _pausedPeers.Contains(key);
    }

    public void SetPeerPaused(string didHex, bool paused)
    {
        if (Normalise(didHex) is not { } key) return;
        lock (_lock)
        {
            var changed = paused ? Add(key) : _pausedPeers.Remove(key);
            if (changed) Persist();
        }

        bool Add(string k)
        {
            if (_pausedPeers.Contains(k)) return false;
            _pausedPeers.Add(k);
            return true;
        }
    }

    /// Untrusted devices the user has hidden from the peer list, so an
    /// office full of strangers' machines doesn't fill the popup. Hiding is
    /// a display preference only: it does not touch trust, and a hidden
    /// device that is later trusted from the other side still has to be
    /// unhidden and trusted here to connect.
    public IReadOnlyList<HiddenPeer> Hidden
    {
        get { lock (_lock) return _hidden.ToList(); }
    }

    public bool IsHidden(string didHex)
    {
        if (Normalise(didHex) is not { } key) return false;
        lock (_lock) return _hidden.Exists(h => h.DidHex == key);
    }

    public void Hide(string didHex, string name)
    {
        if (Normalise(didHex) is not { } key) return;
        lock (_lock)
        {
            if (_hidden.Exists(h => h.DidHex == key)) return;
            _hidden.Add(new HiddenPeer(key, string.IsNullOrWhiteSpace(name)
                ? key[..Math.Min(8, key.Length)] : name));
            Persist();
        }
    }

    public void Unhide(string didHex)
    {
        if (Normalise(didHex) is not { } key) return;
        lock (_lock)
        {
            if (_hidden.RemoveAll(h => h.DidHex == key) == 0) return;
            Persist();
        }
    }

    /// Forget every preference at once — excluded apps, per-peer pauses and
    /// hidden devices — returning the file to its first-run state. Trust
    /// lives in TrustStore and is the caller's to clear alongside this.
    public void ResetAll()
    {
        lock (_lock)
        {
            _excluded.Clear();
            _pausedPeers.Clear();
            _hidden.Clear();
            Persist();
        }
    }

    /// DIDs are compared lowercase; blank ones are not a peer.
    private static string? Normalise(string? didHex) =>
        string.IsNullOrWhiteSpace(didHex) ? null : didHex.Trim().ToLowerInvariant();

    /// Caller holds _lock.
    private void Persist()
    {
        try
        {
            var model = new FileModel
            {
                Version = CurrentVersion,
                ExcludedApps = _excluded.Select(e => new FileModel.Entry
                {
                    Kind = e.Kind == AppKind.Exe ? "exe" : "package",
                    Key = e.Key,
                    Name = e.DisplayName,
                    Path = e.Path,
                }).ToList(),
                PausedPeers = _pausedPeers.ToList(),
                HiddenPeers = _hidden.Select(h => new FileModel.HiddenEntry
                {
                    Did = h.DidHex,
                    Name = h.Name,
                }).ToList(),
            };

            var dir = System.IO.Path.GetDirectoryName(_path);
            if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

            // Write-then-move so a crash mid-write cannot leave a corrupt file.
            var tmp = _path + ".tmp";
            File.WriteAllText(tmp, JsonSerializer.Serialize(model, JsonOptions));
            File.Move(tmp, _path, overwrite: true);
        }
        catch (Exception ex)
        {
            // In-memory list keeps working for this session.
            Security.Log.Write($"AppSettings: could not write {_path}: {ex.GetType().Name}");
        }
    }

    private static bool TryParse(FileModel.Entry e, out AppIdentity identity)
    {
        identity = null!;
        if (string.IsNullOrWhiteSpace(e.Key)) return false;

        AppKind kind;
        switch (e.Kind?.ToLowerInvariant())
        {
            case "exe": kind = AppKind.Exe; break;
            case "package": kind = AppKind.Package; break;
            default: return false;   // e.g. macOS "bundle" entries
        }

        identity = new AppIdentity(kind, e.Key!, e.Name ?? e.Key!, e.Path);
        return true;
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private sealed class FileModel
    {
        public int Version { get; set; } = CurrentVersion;
        public List<Entry> ExcludedApps { get; set; } = new();
        public List<string> PausedPeers { get; set; } = new();
        public List<HiddenEntry> HiddenPeers { get; set; } = new();

        public sealed class Entry
        {
            public string? Kind { get; set; }
            public string? Key { get; set; }
            public string? Name { get; set; }
            public string? Path { get; set; }
        }

        public sealed class HiddenEntry
        {
            public string? Did { get; set; }
            public string? Name { get; set; }
        }
    }
}

File: clipsync-mac/Sources/ClipSync/Settings/AppSettings.swift
import Foundation

/// User preferences, stored as plain JSON at
/// `~/Library/Application Support/ClipSync/settings.json`, alongside
/// `trust.plist`. Same schema as the Windows side (`excludedApps` +
/// `pausedPeers`); macOS entries are `kind: "bundle"` keyed by bundle
/// identifier, and Windows' `exe`/`package` entries are ignored on load
/// rather than treated as errors.
///
/// Not a secret and deliberately hand-editable. A corrupt file degrades
/// to defaults rather than throwing, matching TrustStore.load(). Writes
/// are atomic (write-then-move) so a crash mid-write cannot leave a torn
/// file that silently vanishes the user's exclusions on next launch.
final class AppSettings {
    private static let currentVersion = 1

    private let url: URL
    private var excludedList: [AppIdentity]
    private var pausedList: [String]        // lowercase DID hex, de-duplicated
    private let lock = NSLock()

    private init(url: URL, excluded: [AppIdentity], pausedPeers: [String]) {
        self.url = url
        self.excludedList = excluded
        self.pausedList = pausedPeers
    }

    static func defaultURL() -> URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory,
                                                  in: .userDomainMask).first!
        return appSupport
            .appendingPathComponent("ClipSync", isDirectory: true)
            .appendingPathComponent("settings.json")
    }

    static func load() -> AppSettings { load(url: defaultURL()) }

    static func load(url: URL) -> AppSettings {
        var excluded: [AppIdentity] = []
        var paused: [String] = []
        if FileManager.default.fileExists(atPath: url.path) {
            do {
                let data = try Data(contentsOf: url)
                let model = try JSONDecoder().decode(FileModel.self, from: data)
                for e in model.excludedApps ?? [] {
                    // Unknown kinds (Windows' "exe"/"package", or anything a
                    // future version invents) are skipped, not errors.
                    guard e.kind?.lowercased() == AppIdentity.Kind.bundle.rawValue,
                          let key = e.key,
                          let id = AppIdentity(kind: .bundle, key: key,
                                               displayName: e.name ?? key, path: e.path)
                    else { continue }
                    if !excluded.contains(id) { excluded.append(id) }
                }
                for did in model.pausedPeers ?? [] {
                    // Normalise to lowercase, drop blanks, de-duplicate. An
                    // entry naming a peer this device has never met is kept
                    // deliberately, so pausing a machine, forgetting it and
                    // meeting it again does not silently un-pause it.
                    guard let key = AppSettings.normaliseDid(did) else { continue }
                    if !paused.contains(key) { paused.append(key) }
                }
            } catch {
                // Corrupt or unreadable: start empty rather than crashing.
                NSLog("AppSettings: could not read %@: %@; using defaults",
                      url.path, String(describing: error))
                excluded = []
                paused = []
            }
        }
        return AppSettings(url: url, excluded: excluded, pausedPeers: paused)
    }

    // MARK: Excluded apps

    var excluded: [AppIdentity] {
        lock.lock(); defer { lock.unlock() }
        return excludedList
    }

    func isExcluded(_ app: AppIdentity) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return excludedList.contains(app)
    }

    func add(_ app: AppIdentity) {
        lock.lock(); defer { lock.unlock() }
        guard !excludedList.contains(app) else { return }
        excludedList.append(app)
        persistLocked()
    }

    func remove(_ app: AppIdentity) {
        lock.lock(); defer { lock.unlock() }
        let before = excludedList.count
        excludedList.removeAll { $0 == app }
        guard excludedList.count != before else { return }
        persistLocked()
    }

    // MARK: Paused peers

    /// Peers this device does not send to, by lowercase DID hex.
    ///
    /// Only the per-peer pause lives here. A global pause is deliberately
    /// not persisted: it means "not right now", and a restart that
    /// silently left syncing off would be a bad surprise.
    var pausedPeers: [String] {
        lock.lock(); defer { lock.unlock() }
        return pausedList
    }

    func isPeerPaused(_ didHex: String) -> Bool {
        guard let key = AppSettings.normaliseDid(didHex) else { return false }
        lock.lock(); defer { lock.unlock() }
        return pausedList.contains(key)
    }

    func setPeerPaused(_ didHex: String, paused: Bool) {
        guard let key = AppSettings.normaliseDid(didHex) else { return }
        lock.lock(); defer { lock.unlock() }
        if paused {
            guard !pausedList.contains(key) else { return }
            pausedList.append(key)
        } else {
            let before = pausedList.count
            pausedList.removeAll { $0 == key }
            guard pausedList.count != before else { return }
        }
        persistLocked()
    }

    /// DIDs compare lowercase; a blank one is not a peer.
    static func normaliseDid(_ didHex: String?) -> String? {
        guard let didHex else { return nil }
        let trimmed = didHex.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return trimmed.isEmpty ? nil : trimmed
    }

    // MARK: Persistence

    /// Caller holds `lock`. Writes the whole model — both features share
    /// this file, and writing everything each time is what guarantees
    /// neither clobbers the other.
    private func persistLocked() {
        let model = FileModel(
            version: AppSettings.currentVersion,
            excludedApps: excludedList.map {
                FileModel.Entry(kind: $0.kind.rawValue, key: $0.key,
                                name: $0.displayName, path: $0.path)
            },
            pausedPeers: pausedList
        )
        do {
            let dir = url.deletingLastPathComponent()
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(model)
            // .atomic is write-to-temp-then-rename, the same guarantee the
            // Windows side gets from File.Move.
            try data.write(to: url, options: .atomic)
        } catch {
            // In-memory state keeps working for this session.
            NSLog("AppSettings: could not write %@: %@", url.path, String(describing: error))
        }
    }

    /// The on-disk shape, shared with Windows. Field names are camelCase
    /// to match; everything optional so a partial or foreign file decodes.
    private struct FileModel: Codable {
        var version: Int?       // optional on decode: a file missing it still loads
        var excludedApps: [Entry]?
        var pausedPeers: [String]?

        init(version: Int, excludedApps: [Entry], pausedPeers: [String]) {
            self.version = version
            self.excludedApps = excludedApps
            self.pausedPeers = pausedPeers
        }

        struct Entry: Codable {
            var kind: String?
            var key: String?
            var name: String?
            var path: String?
        }
    }
}
